docs-internal/multipart.md

```diff
@@ -0,0 +1,328 @@
+# Helidon Multipart Support Proposal
+
+Provide an API for the reactive webserver that enables processing multipart requests and returning multipart responses.
+
+## Proposal
+
+Support 2 models: buffered and reactive.
+
+With the buffered model the request content is fully decoded in memory and all the parts are handed-over at the same time.
+This model enables simple use-cases with small size parts.
+
+With the reactive model the request content is decoded for each chunk and the user can subscribe to the stream of chunks for each part.
+This model enables advanced streaming use-cases with large size parts.
+
+Supporting multipart in the webserver means adding support for a new media type. This is done by adding a new entity type with readers and writers.
+
+Each of the 2 models above will be represented by different entity types: `BufferedMultiPart` and `MultiPart`.
+
+```java
+// buffered model
+request.content().as(BufferedMultiPart.class).thenAccept(/* ... */);
+```
+
+```java
+// reactive model
+request.content().as(MultiPart.class).thenAccept(/* ... */);
```

Comment:
The basic idea is buffered / non buffered ; I tried to disambiguate that in the document by calling it the "reactive model".

The reason for not having BufferedMultiPart extend MultiPart is mostly to not have the reactive methods available from a BufferedMultiPart entity and to have a different builder defined on the interface (cannot overload static).

Assuming that we are okay stripping out the "reactive" methods from BufferedMultiPart, there have nothing in common except perhaps the builder stuff.

Reply:
I understand why, in their current forms, `BufferedMultiPart` does not and should not extend `MultiPart`. I'm just saying that the naming is confusing because someone who has not read the document just looking at interface names might reasonably conclude that `BufferedMultiPart` extends `MultiPart.`

Reply:
Roger that. I'm having a hard time with `ReactiveMultiPart`, do you have other suggestions for naming the non buffered entity ?

Reply:
Perhaps `MultiPart` and `StreamingMultipart`?

Reply:
Given (I think) that neither interface will extend the other, what about `BufferedMultiPart` and `StreamingMultiPart`? It could still be slightly confusing to use `MultiPart` for buffered and `StreamingMultiPart` for reactive because (IMO) some people might expect the more specifically-named interface to extend the more generally-named one.

Reply:
In the sake of making the simple use-cases simple we could have MultiPart and StreamingMultiPart like Santiago suggested.

Having BufferedMultiPart and StreamingMultiPart is very explicit but this is probably not useful to know for the user implementing a simple use-cases.